File: projects/06/assembler/parser.py
"""Breaks each assembly command into its underlying components
(fields and symbols).

Public interface
has_more_commands()
advance()
command_type()
symbol()
dest()
comp()
jmp()
"""

class Parser:
    A_COMMAND = 0
    C_COMMAND = 1
    L_COMMAND = 2

    def __init__(self, input_fname):
        """Constructor.
        
        Arguments
        input_fname (string) - Name of input file containing assembly code.
        """
        self._fin = open(input_fname, "r")

    def has_more_commands(self):
        """Checks to see if more commands are available.
        
        Returns
        True if there are more commands, False if not.
        """
        line = self._fin.readline()
        while line != "":
            commands = line.split()
            # Ignore blank lines and comments.
            if len(commands) > 0 and commands[0][0] != '/':
                self._command = commands[0]
                return True
            line = self._fin.readline()
        self._fin.close()
        return False

    def advance(self):
        """Breaks down the command into its constituent parts.
        """
        if self._command[0] == '@':
            self._command_type = Parser.A_COMMAND
            self._parse_a_command()
        elif self._command[1] == '(':
            self._command_type = Parser.L_COMMAND
            self._parse_l_command()
        else:
            self._command_type = Parser.C_COMMAND
            self._parse_c_command()

    def command_type(self):
        return self._command_type

    def symbol(self):
        return self._symbol

    def dest(self):
        return self._dest

    def comp(self):
        return self._comp

    def jmp(self):
        return self._jmp

    def _parse_a_command(self):
        # @symbol
        self._symbol = self._command[1:]

    def _parse_l_command(self):
        # (symbol)
        self._symbol = self._command[1:-1]

    def _parse_c_command(self):
        # dest=comp;jmp
        # Either dest or jmp fields may be empty.
        # If dest is empty the "=" is ommited.
        # If jmp is empty the ";" is ommited.
        comp_start = self._command.find('=')
        if comp_start != -1:
            self._dest = self._command[:comp_start]
            comp_start += 1
        else:
            self._dest = ""
            comp_start = 0
        comp_end = self._command.rfind(';')
        if comp_end != -1:
            self._comp = self._command[comp_start:comp_end]
            self._jmp = self._command[comp_end + 1:]
        else:
            self._comp = self._command[comp_start:]
            self._jmp = ""

File: projects/07/translator/src/translator/Parser.java
package translator;

import java.io.BufferedReader;
import java.io.Closeable;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;

/**
 * Handles parsing of a single VM file.
 * @author kujawk
 *
 */
class Parser implements Closeable
{
    private final BufferedReader reader;
    private String[] tokens;
    private String command;
    private String arg1;
    private String arg2;
    
    /**
     * Creates a parser for VM file.
     * @param inputFileName
     * @throws FileNotFoundException If the file is not found. 
     */
    Parser(File inputFile) throws FileNotFoundException
    {
        reader = new BufferedReader(new FileReader(inputFile));
    }
    
    /**
     * Checks to see if more commands) are available.    
     * @return True if there are more commands, false if not.
     * @throws IOException If a problem occurred reading from the file.
     */
    boolean hasMoreCommands() throws IOException
    {
        try
        {
            String line = reader.readLine();
            while (line != null)
            {
                tokens = line.split("\\s+");
                // Ignore blank lines and comments.
                if (tokens.length > 0 && !tokens[0].isEmpty() && tokens[0].charAt(0) != '/')
                {
                	return true;
                }
                line = reader.readLine();
            }
            reader.close();
            return false;
        }
        catch (IOException e)
        {
            reader.close();
            throw e;
        }
    }
    
    /**
     * Breaks down the current command into its constituent parts.
     */
    void advance()
    {
        command = tokens[0];
        arg1 = tokens.length > 1 ? tokens[1] : null;
        arg2 = tokens.length > 2 ? tokens[2] : null;
    }
    
    /**
     * @return Returns the current command.
     */
    String command()
    {
        return command;
    }
    
    /**
     * Returns the first argument of the current command, if the command 
     * has one.
     * @return The first argument of the current command.
     */
    String arg1()
    {
        return arg1;
    }
    
     /**
     * Returns the second argument of the current command, if the command 
     * has one.
     * @return The second argument of the current command.
     */
    String arg2()
    {
        return arg2;
    }

	@Override
	public void close() throws IOException {
		// TODO Auto-generated method stub
	}
}
